Source/Particles/Collision/BackgroundStopping/BackgroundStopping.cpp

```diff
@@ -0,0 +1,266 @@
+    if (background_type_str == "electrons") {
+        m_background_type = BackgroundStoppingType::ELECTRONS;
+    } else if (background_type_str == "ions") {
+        m_background_type = BackgroundStoppingType::IONS;
+    } else {
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(false, "background_type must be either electrons or ions");
+    }
+
+    amrex::Real background_density;
+    std::string background_density_str;
+    if (queryWithParser(pp_collision_name, "background_density", background_density)) {
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(background_density > 0,
+                 "For background stopping, the background density must be greater than 0");
+        m_background_density_parser = makeParser(std::to_string(background_density), {"x", "y", "z", "t"});
+    } else if (pp_collision_name.query("background_density(x,y,z,t)", background_density_str)) {
+        m_background_density_parser = makeParser(background_density_str, {"x", "y", "z", "t"});
+    } else {
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(false,
+                 "For background stopping, the background density must be specified.");
+    }
+
+    amrex::Real background_temperature;
+    std::string background_temperature_str;
+    if (queryWithParser(pp_collision_name, "background_temperature", background_temperature)) {
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(background_temperature > 0,
+                 "For background stopping, the background temperature must be greater than 0");
+        m_background_temperature_parser = makeParser(std::to_string(background_temperature), {"x", "y", "z", "t"});
+    } else if (pp_collision_name.query("background_temperature(x,y,z,t)", background_temperature_str)) {
+        m_background_temperature_parser = makeParser(background_temperature_str, {"x", "y", "z", "t"});
+    } else {
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(false,
+                 "For background stopping, the background temperature must be specified.");
+    }
+
+    m_background_density_func = m_background_density_parser.compile<4>();
+    m_background_temperature_func = m_background_temperature_parser.compile<4>();
+
+    if (m_background_type == BackgroundStoppingType::ELECTRONS) {
+        m_background_mass = PhysConst::m_e;
+        queryWithParser(pp_collision_name, "background_mass", m_background_mass);
+    } else if (m_background_type == BackgroundStoppingType::IONS) {
+        getWithParser(pp_collision_name, "background_mass", m_background_mass);
+        getWithParser(pp_collision_name, "background_charge_state", m_background_charge_state);
+    }
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(m_background_mass > 0,
+             "For background stopping, the background mass must be greater than 0");
+
+}
+
+void
+BackgroundStopping::doCollisions (amrex::Real cur_time, MultiParticleContainer* mypc)
+{
+    WARPX_PROFILE("BackgroundStopping::doCollisions()");
+    using namespace amrex::literals;
+
+    const amrex::Real dt = WarpX::GetInstance().getdt(0);
```

Comment:
Based on the discussion during the 3/1/22 development meeting about `dt` (possibly) changing on different refinement levels, should a level specific `dt` be passed to `doBackgroundStoppingOnElectronsWithinTile` and `doBackgroundStoppingOnIonsWithinTile` below?

Reply:
Yes, thanks. I'll fix

Reply:
Thanks for the discussion on this. If the collisions are only applied on the lowest level should the loop over refinement levels (line 94) still be done? I'm not fully understanding what happens on the different levels during that loop then.

Reply:
The loop over all levels is still needed here so that particles at all levels are operated on, since particles are stored relative to the highest level block at their location (is that correct?).